### PR TITLE
8290451: Incorrect result when switching to C2 OSR compilation from C1

### PR DESCRIPTION
**Pull request by rwestrel**
In BadStateAtLongCmp.test(), the field increment is between the lcmp
and iflt bytecodes that implement the loop exit condition (handcrafted
bytecodes required as such a sequence is not generated by javac). When
compiled by C1, lcmp+iflt is canonicalized to a single If and it
captures the state of the lcmp bytecode. Tiered compilation attaches
the state to a deoptimization call to switch to an OSR
compilation. That state causes the field increment to be reexecuted
when the switch to a C2 method occurs at the backbranch. Proposed fix
is to use the state at the iflt instead.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8290451](https://bugs.openjdk.org/browse/JDK-8290451): Incorrect result when switching to C2 OSR compilation from C1

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9938/head:pull/9938` \
`$ git checkout pull/9938`

Update a local copy of the PR: \
`$ git checkout pull/9938` \
`$ git pull https://git.openjdk.org/jdk pull/9938/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9938`

View PR using the GUI difftool: \
`$ git pr show -t 9938`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9938.diff">https://git.openjdk.org/jdk/pull/9938.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/c1/BadStateAtLongCmp.jasm [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+super public class BadStateAtLongCmp
+	version 61:0
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c1/BadStateAtLongCmp.jasm:25]: Please use an older bytecode version to simplify backports.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

File: test/hotspot/jtreg/compiler/c1/BadStateAtLongCmp.jasm (hunk 1 above)
Comment (rwestrel) [reply test/hotspot/jtreg/compiler/c1/BadStateAtLongCmp.jasm]: Done (with the class file version for jdk 8 now)

**Comment by rwestrel** [comment]
> Looks good to me.

Thanks for the review.

**Review by vnkozlov — APPROVED** [review]
Ok.

**Review by TobiHartmann — APPROVED** [review]

**Comment by rwestrel** [comment]
@vnkozlov thanks for the review

**Pull request closed without merging**